### ISSUE TITLE
Dart_Null inaccessible from dart_api_dl ?

### ISSUE THREAD

**Issue by natinusala** [dart-lang/sdk#51794]
Is there a technical reason why `Dart_Null()` isn't accessible from `dart_api_dl.h`, or is it just that it needs to be added? (would FFI support it?) We have `Dart_IsNull()` already.

* Dart SDK Version (`dart --version`) : 2.19.2
* Whether you are using Windows, MacOSX, or Linux (if applicable) : WSL Ubuntu 22.something

**Comment by faizan003** [comment]
The Dart_Null() function is not currently accessible from dart_api_dl.h because it is not part of the public Dart API. This function is used internally by the Dart runtime to represent the null value in certain contexts, but it is not intended to be used directly by external code.

Instead, external code should use the Dart_IsNull() function to check if a Dart_Handle value represents the null value. This function is part of the public Dart API and is safe to use in external code.

Regarding FFI support, it's unlikely that Dart_Null() would be directly supported by FFI, as it's not part of the public API. However, FFI does provide mechanisms for working with native null pointers, such as the nullptr value in C++, or the NULL macro in C. If you need to pass a null value to a Dart function via FFI, you can use one of these mechanisms to represent the null value on the native side, and then use the Dart_Null() function to create a Dart_Handle value representing null inside the Dart function.

**Comment by natinusala** [comment]
Okay I understand, I thought it was part of the public API.

What do you mean exactky by `use the Dart_Null() function to create a Dart_Handle value representing null inside the Dart function.`? 

How can I use `Dart_Null()` if it's not part of the public API? Should I be calling it from Dart?

I tried to pass `nullptr` to FFI before and IIRC it crashed the Dart program (with an error similar to "invalid Handle expected Instance"). Was I missing something?

**Comment by faizan003** [comment]
Yes, you are correct that Dart_Null() is not part of the public API, so you should not call it from your FFI code. Instead, you can create a null handle inside your Dart code using Dart_Null() and pass it as a parameter to your FFI function. Then, in your FFI code, you can check if the received handle is null by using Dart_IsNull().

Regarding passing nullptr to FFI, you should not do that because it is not a valid Dart handle value. When you pass nullptr to a Dart FFI function, the Dart VM will not be able to find the corresponding Dart handle value for the null pointer and will throw an error. Instead, you can pass a valid Dart handle representing null, which can be created using Dart_Null() in your Dart code.

**Comment by natinusala** [comment]
Is this an AI generated answer? FFS how do I report that

**Comment by mraleph** [comment]
@natinusala yeah it is. I have hidden or deleted comments across all issues where this person "commented'.

**Comment by a-siva** [comment]
//cc @dcharkes 
Should Dart_Null be added to the list ?

**Comment by dcharkes** [comment]
We don't need to, you can pass values such as `null`, `true`, `false` via FFI as `Dart_Handle`.

https://dart-review.googlesource.com/c/sdk/+/293541

I think it's preferable to only expose API functions that are really needed.

**Comment by dcharkes** [comment]
@natinusala does passing the `null` value via an FFI call work for you? Or do you have a use case for which `Dart_Null()` would be more suited?

**Comment by a-siva** [comment]
based on last comment from @natinusala closing issue.

**Comment by natinusala** [comment]
@dcharkes @a-siva sorry I haven't got time to try this again until today. 

What I need is the ability to return a `null` handle in a C -> Dart callback.

I have a C function pointer that is the result of a call to `Pointer.fromFunction` in Dart. That function takes a `Dart_Handle` as a parameter, and I'd like it to be converted to an `Object?` in the Dart implementation.

That works, but I cannot seem to create a "null handle" in C to give back to Dart. When I call the function with a normal handle it works, but if I replace the value with the C `NULL`, the Dart VM seems to crash (kinda expected).

I expected (the nonexistent) `Dart_Null_DL()`  to do exactly that, allow C code to create a "null handle" back to Dart in a callback (or as a return value...).

**Comment by dcharkes** [comment]
> That works, but I cannot seem to create a "null handle" in C

Do you have a Dart->C call that calls that C->Dart callback? Then you could construct the null handle by passing it from Dart.

If not, we should add `Dart_Null_DL()` to `dart_api_dl.h`.

**Comment by dcharkes** [comment]
Try without FFIgen.

**Comment by natinusala** [comment]
@dcharkes it works! I'll close the issue again
